### PR TITLE
Add recipe for dlib 19.19 

### PR DESCRIPTION
**Pull request by Morwenn**
Specify library name and version:  **dlib/19.19**

Here is a simple recipe for dlib, with still a few pain points:
- I don't handle the libjpeg/libjpeg-turbo option (but I believe that we need to discuss the libjpeg/libjpeg-rubo/mozjpeg issues in CCI recipes globally at some point)
- The giflib option is defaulted to `False` because it apparently doesn't build on MSVC outside of a Windows subsystem: I don't exclude it, but it's not guaranteed to build
- dlib embeds pybind11 and gives no option to replace it with a user-provided one as far as I know
- I've got `["auto", True, False]` lists for SIMD options which mirror the ones in the CMakeLists.txt, I'm not sure whether those are a good idea or what would be the preferred alternative there
- I do use `tools.collect_libs` because dlib seems to generate a single library file but its name depends on the settings (for example mine is called `dlib19.19.0_release_64bit_msvc1916.lib`)

The example in `test_package.cpp` is from the dlib repository itself - all of their examples are apparently public domain.

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/wiki#how-to-submit-a-pull-request) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/dlib/all/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
+import os
+
+from conans import ConanFile, CMake, tools
+
+
+class DlibConan(ConanFile):
+    name = "dlib"
+    description = "A toolkit for making real world machine learning and data analysis applications"
+    topics = "conan", "dlib"
```
Comment (uilianries) [inline recipes/dlib/all/conanfile.py:9]:
```suggestion
    topics = ("machine-learning", "deep-learning", "computer-vision")
```

File: recipes/dlib/all/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+            os.path.join("lib", "cmake"),
+            os.path.join("lib", "pkgconfig"),
+            os.path.join("include", "dlib", "cmake_utils"),
+            os.path.join("include", "dlib", "external", "pybind11", "tools")
+        ]:
+            tools.rmdir(os.path.join(self.package_folder, dir_to_remove))
+
+    def package_info(self):
+        # There is a single library whose name depends on settings
+        self.cpp_info.libs = tools.collect_libs(self)
```
Comment (uilianries) [inline recipes/dlib/all/conanfile.py:113]:
```suggestion
        self.cpp_info.libs = tools.collect_libs(self)
        if self.settings.os == "Linux":
            self.system_info.libs = ["pthread"]
```

**Comment by Morwenn** [comment]
We're triggering that error for Visual Studio:

```
CMake Error at source_subfolder/dlib/CMakeLists.txt:108 (message):
  Building dlib as a standalone dll is not supported when using Visual
  Studio.  You are highly encouraged to use static linking instead.  See
  https://github.com/davisking/dlib/issues/1483 for a discussion.
```

Should that be reflected through a configuration error, or should be remove the `shared` option with Visual Studio? I'm not sure which is the right solution here.

File: recipes/dlib/all/conanfile.py (hunk 2 above)
Comment (Morwenn) [reply recipes/dlib/all/conanfile.py:113]: Done.

File: recipes/dlib/all/conanfile.py (hunk 1 above)
Comment (Morwenn) [reply recipes/dlib/all/conanfile.py]: Done.

**Comment by uilianries** [comment]
Yes, drop it by raising ConanInvalidConfiguration.

File: recipes/dlib/all/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        # we want to be able to build it as a shared library too
+        self._cmake.definitions["DLIB_IN_PROJECT_BUILD"] = "OFF"
+
+        # Configure external dependencies
+        self._cmake.definitions["DLIB_GIF_SUPPORT"] = self.options.with_gif
+        self._cmake.definitions["DLIB_JPEG_SUPPORT"] = self.options.with_jpeg
+        self._cmake.definitions["DLIB_PNG_SUPPORT"] = self.options.with_png
+
+        # Configure SIMD options
+        if self.options.with_sse2 != "auto":
```
Comment (SSE4) [inline recipes/dlib/all/conanfile.py:87]: these options (sse2, avx) are architecture specific. 
remove them if arch is neither x86 nor x86_64

Comment (Morwenn) [reply recipes/dlib/all/conanfile.py]: It should be addressed now :)

**Pull request merged**